src/auth/github-auth.service.ts

```diff
@@ -0,0 +1,121 @@
+import {
+    Injectable,
+    Inject,
+    InternalServerErrorException,
+    NotFoundException,
+} from "@nestjs/common";
+import {
+    AuthUserResult,
+    IAuthProvider,
+} from "../global/interfaces/oauth.interface";
+import { PrismaService } from "../prisma/prisma.service";
+import { GithubUser } from "../global/types/auth.types";
+import { generatePasswordHash } from "../global/auth/utils";
+import { OAuthConfig } from "@/config/Oauth/oauthConfig.interface";
+
+@Injectable()
+export class GithubAuthService implements IAuthProvider {
+    constructor(
+        private prisma: PrismaService,
+        @Inject("OAuth-Config") private oAuthConfig: OAuthConfig,
+    ) {}
+    async validateUser(user: GithubUser) {
+        const userInDb = await this.prisma.findUserByOAuthId(
+            "github",
+            user.githubId,
+        );
+
+        if (userInDb)
+            return {
+                id: userInDb.userId,
+                email: user.email,
+            };
+        return this.createUser(user);
+    }
+
+    async createUser(user: GithubUser): Promise<AuthUserResult> {
+        // generate a random password and not tell them so they can't login, but they will be able to reset password,
+        // and will be able to login with this in future,
+        // or maybe the app will prompt user the input the password, exact oauth flow is to be determined
+
+        // this should not happen when "user:email" is in the scope
+        if (!user.email)
+            throw new InternalServerErrorException(
+                "[github-auth.service]: Cannot get email from github to create a new Chingu account",
+            );
+
+        const existingUser = await this.findUserByEmails([
```

Comment:
Is it possible that this could return `null`?

Reply:
yes according to the Prisma API https://www.prisma.io/docs/orm/reference/prisma-client-reference#findfirst